### DISCUSSION TITLE
What are directives

### DISCUSSION THREAD

**Discussion by tomjn** [WordPress/gutenberg#53040] [Interactivity API]
I've come here from the Make WP post, but both that post, this repo, and all the issues have this problem.

The goal of the project to make building interactive blocks easier is well stated, but none of the experiments or attempts to achieve that are explained, and they're all referred to by their terminology which is never explained. 

For example, neither the Make post, the readme, or the main tracking issue ever says what directives are. This is the closest anybody gets to explaining what directives are:

> The goal is to create an installable plugin that adds a set of basic directives and client-side transitions to any site.

_But what are directives?_ And thanks to anybody who tries to explain in this issue, but it needs adding to the readme.

The repo is named after hydration experiments, and that gives some clues based on some articles I read a long time ago about React server side rendering and rehydrating components in the browser, but that too is unclear in the context of this project and blocks. I might interpret this as allowing you to build both the editor and frontend interfaces in React, but what I've read elsewhere doesn't fit this interpretation.

**My fear is that this project is doing great work but has failed to explain what it's doing, and couched itself in opaque terminology.** There are no examples or use cases in the readme, no screenshots or videos, no glossary, etc. I should have been able to tell what these things were within 30 seconds via the readme.

**Comment by justintadlock** [comment]
I also replied to this [in the comments](https://make.wordpress.org/core/2023/01/11/update-on-the-work-to-make-building-interactive-blocks-easier/#comment-44285) on the Make Core post.  We definitely need more thorough explanations of what this project is doing, more so that we can bring other developers along for the journey.

Directives are special HTML attributes that allow developers to add interactivity to their markup.  A simple example might be using a `wp-show` directive:

```html
<div wp-show="state.shouldShow">content</div>
```

In this example, the inner content would only show if `state.shouldShow` is `true` or not if `false`.

This is similar to other JavaScript frameworks like Angular, Vue, and Alpine but will have some differences that will be explained when the proposal is made public.

**Comment by tomjn** [comment]
I'm deeply saddened that this got turned into a discussion, **because this is a bug, and actionable steps are needed.**

Where is the bug: the readme file
What is wrong: it does not explain the basic purpose of the projects, or what the terms it uses mean at their simplest level
What is needed: for that to be added to the readme

Documentation and messaging can have problem too not just code, and they can be fixed too just like code.

@justintadlock one of the things I desperately wanted to avoid when I left my issue, was it being twisted from its intention and turned from a "this project isn't clear" issue, into a "Tom doesn't understand and wants a personal explanation" issue. I trust that I'll figure it out on my own and I had a pretty fair idea already when I raised the issue, but not everybody has the time, familiarity, or knowledge to do that.

**Reply by Poliuk** [reply]
Hey Tom, thanks again for your feedback. You are entirely correct; the current status of the repo could be better regarding how things are explained, and we are aware of it. 

Improving the documentation of this project is the next step in the process. As stated in the [Make WP Post](https://make.wordpress.org/core/2023/01/11/update-on-the-work-to-make-building-interactive-blocks-easier/), the contributors who have been working on this repo are preparing an initial proposal that aims to give full context and detailed explanations about what this is all about. The readme will be improved, a demo will be created, a new blog post in Make Core will be written, and a technical Design Doc will be available... 

We are sorry about your issue being turned into a discussion. We were not planning on ignoring the fact that you are pointing out. So if you feel like this should be an issue, feel free to re-open it, I'm sure the contributors working on this will be happy to provide visibility on it as soon as they start working on the different pieces of documentation.
